src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs

```diff
@@ -117,5 +118,107 @@ internal static int IndexOfStraddling(this ReadOnlySpan<byte> first, IReadOnlyMe
 
             return -1;
         }
+
+        static readonly int s_longSize = Vector<ulong>.Count;
+        static readonly int s_byteSize = Vector<byte>.Count;
+
+        public static int IndexOfVectorized(this Span<byte> buffer, byte value)
+        {
+            Debug.Assert(s_longSize == 4 || s_longSize == 2);
+
+            var byteSize = s_byteSize;
+
+            if (buffer.Length < byteSize * 2 || !Vector.IsHardwareAccelerated) return buffer.IndexOf(value);
```

Comment:
Test `!Vector.IsHardwareAccelerated` first so the Jit eliminates everything in the function. Might be worth adding a definitely inlined indirection shim?

```csharp
public static int IndexOfVectorized(this Span<byte> buffer, byte value)
{
    if (Vector.IsHardwareAccelerated && buffer.Length >= Vector<byte>.Count)
    {
        return buffer.IndexOfVectorizedImpl(value);
    }

    return buffer.IndexOf(value);
}
```